### PR TITLE
[cdac] Implement ISOSDacInterface::GetObjectData getting COM RCW/CCW

### PR DESCRIPTION
**Pull request by elinor-fung**
- Add `SyncTableEntry`, `SyncBlock`, `InteropSyncBlockInfo` data descriptors, `SyncTableEntries` global
- Get the sync block corresponding to a given object address in `Object` contract
- Add `GetComData` to `Object` contract
- Finish implementation of `ISOSDacInterface::GetObjectData` such that it populates the RCW/CCW
- Add test helpers for mocking out sync blocks

Contributes to https://github.com/dotnet/runtime/issues/99302

Manually validated with a COM server and callback that the RCW/CCW were set when expected and matched values returned by DAC.
Ran SOS unit tests from diagnostics repo using private build.

### CODE REVIEW

File: src/coreclr/debug/daccess/request.cpp [hunk 1]
```diff
@@ -71,6 +71,7 @@ TADDR DACGetMethodTableFromObjectPointer(TADDR objAddr, ICorDebugDataTarget * ta
     ULONG32 returned = 0;
     TADDR Value = (TADDR)NULL;
 
+    // Every object has a pointer the methad table at offset 0
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/debug/daccess/request.cpp:74]:
```suggestion
    // Every object has a pointer to its method table at offset 0
```

File: src/coreclr/vm/syncblk.h [hunk 2]
```diff
@@ -969,6 +969,17 @@ class InteropSyncBlockInfo
...
 #endif // FEATURE_OBJCMARSHAL
+
+    template<typename T> friend struct ::cdac_data;
+};
+
+template<>
+struct cdac_data<InteropSyncBlockInfo>
+{
+#ifdef FEATURE_COMINTEROP
+    static constexpr size_t CCW = offsetof(InteropSyncBlockInfo, m_pCCW);
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/syncblk.h:980]: I assume this is only for built-in. Are we going to defer `ComWrappers` support for later? Same for RCW.

File: src/native/managed/cdacreader/src/Contracts/Object_1.cs [hunk 3]
```diff
@@ -11,15 +11,32 @@ namespace Microsoft.Diagnostics.DataContractReader.Contracts;
...
     private readonly byte _objectToMethodTableUnmask;
+    private readonly TargetPointer _stringMethodTable;
+    private readonly TargetPointer _syncTableEntries;
+
+    private static class SyncBlockValue
+    {
+        [Flags]
+        public enum Bits
+        {
+            // Value represents either the hash code ore sync block index (bits 0-25)
```
Comment (AaronRobinsonMSFT) [inline src/native/managed/cdacreader/src/Contracts/Object_1.cs:23]:
```suggestion
            // Value represents either the hash code or sync block index (bits 0-25)
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/coreclr/vm/syncblk.h (hunk 2 above)
Comment (jkoritzinsky) [reply src/coreclr/vm/syncblk.h:980]: What about if we call it "BuiltInComData" to match how our docs and current code talk about built-in COM vs ComWrappers?

Then in the future, the contract to pull data about built-in RCWs and CCWs could be called "BuiltInCom" and the ComWrappers one could be "ComWrappers".

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**